text/0517-io-os-reform.md

```diff
@@ -473,12 +649,229 @@ throughout IO, we can go on to explore the modules in detail.
+    fn read_until(&mut self, byte: u8) -> NonatomicResult<Vec<u8>, Vec<u8>, Self::Err> { ... }
+}
+
+pub trait BufferedReaderExt: BufferedReader {
+    fn lines(&mut self) -> Lines<Self, Self::Err> { ... };
+}
+```
+
+In addition, `read_line` is removed in favor of the `lines` iterator,
+and `read_char` is removed in favor of the `chars` iterator (now on
+`ReaderExt`). These iterators will be changed to yield
+`NonatomicResult` values.
+
+The `BufferedReader`, `BufferedWriter` and `BufferedStream` types stay
+essentially as they are today, except that for streams and writers the
+`into_inner` method yields any errors encountered when flushing,
+together with the remaining data:
+
+```rust
+// If flushing fails, you get the unflushed data back
+fn into_inner(self) -> NonatomicResult<W, Vec<u8>, W::Err>;
+```
+
+#### `MemReader` and `MemWriter`
+[MemReader and MemWriter]: #memreader-and-memwriter
+
+The various in-memory readers and writers available today will be
+consolidated into just `MemReader` and `MemWriter`:
+
+`MemReader` (like today's `BufReader`)
+ - construct from `&[u8]`
+ - implements `Seek`
+
+`MemWriter`
+ - construct freshly, or from a `Vec`
+ - implements `Seek`
+
+Both will allow decomposing into their inner parts, though the exact
+details are left to the implementation.
+
+The rationale for this design is that, if you want to read from a
+`Vec`, it's easy enough to get a slice to read from instead; on the
+other hand, it's rare to want to write into a mutable slice on the
+stack, as opposed to an owned vector. So these two readers and writers
+cover the vast majority of in-memory readers/writers for Rust.
+
+In addition to these, however, we will have the following `impl`s
+directly on slice/vector types:
+
+* `impl Writer for Vec<u8>`
+* `impl Writer for &mut [u8]`
+* `impl Reader for &[u8]`
+
+These `impls` are convenient and efficient, but do not implement `Seek`.
 
 ### The `std::io` facade
 [The std::io facade]: #the-stdio-facade
 
-> To be added in a follow-up PR.
+The `std::io` module will largely be a facade over `core::io`, but it
+will add some functionality that can live only in `std`.
+
+#### `Errors`
+[Errors]: #error
+
+The `IoError` type will be renamed to `std::io::Error`, following our
+[non-prefixing convention](https://github.com/rust-lang/rfcs/pull/356).
+It will remain largely as it is today, but its fields will be made
+private. It may eventually grow a field to track the underlying OS
+error code.
+
+The `IoErrorKind` type will become `std::io::ErrorKind`, and
+`ShortWrite` will be dropped (it is no longer needed with the new
+`Writer` semantics), which should decrease its footprint. The
+`OtherIoError` variant will become `Other` now that `enum`s are
+namespaced.
+
+#### Channel adapters
+[Channel adapters]: #channel-adapters
+
+The `ChanReader` and `ChanWriter` adapters will be kept exactly as they are today.
+
+#### `stdin`, `stdout`, `stderr`
+[stdin, stdout, stderr]: #stdin-stdout-stderr
```

Comment:
I was talking about the `set_stdout` and `set_stderr` functions, rather than global vs thread local buffering.

(This is also just a clarification question, I'm not pushing either way to keep or remove them)